Source/WebKit/WebProcess/GPU/graphics/cocoa/ImageBufferShareableMappedIOSurfaceBitmapBackend.cpp

```diff
@@ -74,15 +74,41 @@ ImageBufferBackendHandle ImageBufferShareableMappedIOSurfaceBitmapBackend::creat
 
 GraphicsContext& ImageBufferShareableMappedIOSurfaceBitmapBackend::context()
 {
-    if (!m_context) {
+    if (m_context) {
+        CGContextRef cgContext = m_context->platformContext();
+        if (m_lock || !cgContext) {
+            // The existing context is a valid context and the IOSurface is locked, or alternatively
+            // the existing context is an invalid context, for some reason we ran into an error previously.
+            return *m_context;
+        }
+
+        // The IOSurface is unlocked for every flush to prepare for external access by the compositor.
+        // Re-lock on first context() request after the external access has ended and new update starts.
+        if (auto lock = m_surface->lock(IOSurface::AccessMode::ReadWrite)) {
+            auto bitsPerComponent = m_surface->bitmapConfiguration().bitsPerComponent;
+            auto bitsPerPixel = bitsPerComponent * 4;
+            auto bytesPerRow = m_surface->bytesPerRow();
+            auto size = m_surface->size();
+            auto success = CGBitmapContextSetData(cgContext, 0, 0, size.width(), size.height(), lock->surfaceBaseAddress(), bitsPerComponent, bitsPerPixel, bytesPerRow);
+            if (success) {
+                m_lock = WTFMove(lock);
+                return *m_context;
+            }
+        }
+        m_context = nullptr;
+    } else {
         auto lockAndContext = m_surface->createBitmapPlatformContext();
         if (lockAndContext) {
             m_lock = WTFMove(lockAndContext->lock);
             m_context = makeUnique<GraphicsContextCG>(lockAndContext->context.get());
-        } else
-            m_context = makeUnique<GraphicsContextCG>(nullptr);
-        applyBaseTransform(*m_context);
+            applyBaseTransform(*m_context);
+            return *m_context;
+        }
     }
+    // For some reason we ran into an error. Construct an invalid context, with current API we must
```

Comment:
Should we RELEASE_LOG here?

Reply:
I suppose. I'm not a fan but maybe that motivates handling the errors properly in the caller (and then moving the release log there, and then repeating the procedure..)